docs/site/Interceptors.md

```diff
@@ -0,0 +1,328 @@
+  /**
+   * Construct a new instance
+   * @param parent Parent context
+   * @param target Target class or object
+   * @param methodName Method name
+   * @param args An array of arguments
+   */
+  constructor(
+    parent: Context,
+    public readonly target: object,
+    public readonly methodName: string,
+    public readonly args: any[],
+  ) {
+    super(parent);
+  }
+}
+```
+
+It's possible for an interceptor to mutate items in the `args` array to pass in
+transformed input to downstream interceptors and the target method.
+
+### Example interceptors
+
+Here are some example interceptor functions:
+
+```ts
+// A sync interceptor to log method calls
+const logSync: Interceptor = (invocationCtx, next) => {
+  console.log('logSync: before-' + invocationCtx.methodName);
+  // Calling `next()` without `await`
+  const result = next();
+  // It's possible that the statement below is executed before downstream
+  // interceptors or the target method finish
+  console.log('logSync: after-' + invocationCtx.methodName);
+  return result;
+};
+
+// An async interceptor to log method calls
+const log: Interceptor = async (invocationCtx, next) => {
+  console.log('log: before-' + invocationCtx.methodName);
+  // Wait until the interceptor/method chain returns
+  const result = await next();
+  console.log('log: after-' + invocationCtx.methodName);
+  return result;
+};
+
+// An interceptor to catch and log errors
+const logError: Interceptor = async (invocationCtx, next) => {
+  console.log('logError: before-' + invocationCtx.methodName);
+  try {
+    const result = await next();
+    console.log('logError: after-' + invocationCtx.methodName);
+    return result;
+  } catch (err) {
+    console.log('logError: error-' + invocationCtx.methodName);
+    throw err;
+  }
+};
+
+// An interceptor to convert `name` arg to upper case
+const convertName: Interceptor = async (invocationCtx, next) => {
+  console.log('convertName: before-' + invocationCtx.methodName);
+  invocationCtx.args[0] = (invocationCtx.args[0] as string).toUpperCase();
+  const result = await next();
+  console.log('convertName: after-' + invocationCtx.methodName);
+  return result;
+};
+```
+
+To leverage dependency injection, a provider class can be defined as the
+interceptor:
+
+```ts
+class NameValidator implements Provider<Interceptor> {
+  constructor(@inject('valid-names') private validNames: string[]) {}
+  value() {
+    return async <T>(
+      invocationCtx: InvocationContext,
+      next: () => ValueOrPromise<T>,
+    ) => {
+      const name = invocationCtx.args[0];
+      if (!this.validNames.includes(name)) {
+        throw new Error(
+          `Name '${name}' is not on the list of '${this.validNames}`,
+        );
+      }
+      return await next();
+    };
+  }
+}
+```
+
+## Apply interceptors
+
+Interceptors form a cascading chain of handlers around the target method
+invocation. We can apply interceptors by decorating methods/classes with
+`@intercept`.
```

Comment:
I find this misleading. `@intercept` is  not replacing the actual method body, but only creating a new metadata entry. IIUC the source code, interceptors are applied only when the method is invoked via `invokeMethodWithInterceptors`.

Please make this fact very clear in the docs.